spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/reactor/ReactorSleuth.java

```diff
@@ -88,39 +90,51 @@ private ReactorSleuth() {
 				return sub;
 			}
 
+			Context context = sub.currentContext();
+
 			if (log.isTraceEnabled()) {
-				log.trace("Spring Context [" + springContext
-						+ "] Creating a scope passing span subscriber with Reactor Context "
-						+ "[" + sub.currentContext() + "] and name [" + name(sub) + "]");
+				log.trace("Spring context [" + springContext + "], Reactor context ["
+						+ context + "], name [" + name(sub) + "]");
+			}
+
+			// Try to get the current trace context bean, lenient when there are problems
+			CurrentTraceContext currentTraceContext = lazyCurrentTraceContext.get();
+			if (currentTraceContext == null) {
+				if (log.isTraceEnabled()) {
+					log.trace("Spring Context [" + springContext
+							+ "] did not return a CurrentTraceContext. Reactor Context is ["
+							+ sub.currentContext() + "] and name is [" + name(sub) + "]");
+				}
+				assert false; // should never happen, but don't break.
+				return sub;
+			}
+
+			TraceContext parent = traceContext(context, currentTraceContext);
+			if (parent == null) {
+				return sub; // no need to scope a null parent
 			}
 
-			return scopePassingSpanSubscription(springContext, sub);
+			if (log.isTraceEnabled()) {
+				log.trace("Creating a scope passing span subscriber with Reactor Context "
+						+ "[" + context + "] and name [" + name(sub) + "]");
+			}
+			return new ScopePassingSpanSubscriber<>(sub, context, currentTraceContext,
+					parent);
 		});
 	}
 
 	static String name(CoreSubscriber<?> sub) {
 		return Scannable.from(sub).name();
 	}
 
-	private static Map<ConfigurableApplicationContext, CurrentTraceContext> CACHE = new ConcurrentHashMap<>();
-
-	static <T> CoreSubscriber<? super T> scopePassingSpanSubscription(
-			ConfigurableApplicationContext springContext, CoreSubscriber<? super T> sub) {
-		CurrentTraceContext currentTraceContext = CACHE.computeIfAbsent(springContext,
```

Comment:
@robotmrv so the lazybean thing eliminates the need for this cache as far as I can tell.